Improved error messages for type mismatches

MekhrubonT opened the pull request:
This PR tries to improve error output for type mismatch during compilation .
Before:
The second has explicit return type and the first doesn't.

Now:
Constructors do not match:
Foo of 'a
is not compatible with:
Foo : int -> int t
The second has explicit return type and the first doesn't.

During variant and record mismatch the type is zoomed and mistaken constructors' or fields' names and types are printed additionally.

Commit:
Printing additional info about types mismatch during compilation error.

Commit:
Added new tests for type mismatch errors

MekhrubonT commented:
Note: this was previously part of #8821 -- the present PR is improved thanks to the review comments over there.

Octachron commented:
Going through the testsuite changes, I feel that in many(74% in fact) cases the error message just becomes more noisy because the information was already here and is only being repeated in a slightly different form. For instance, in

```OCaml
         type ('a, 'b) bar += A of 'b
       is not included in
         type ('a, 'b) bar += A of 'a
       The types for constructor A are not equal.
       Constructors do not match:
         A of 'b
       is not compatible with:
         A of 'a0
       The types are not equal.
```
displaying the constructor separately does not bring anything new to the table.

Contrarily, the new error submessage seems much more useful when the type definition is not available. For instance in
```OCaml
3 | module type B = A with type t = u;; (* fail *)
                           ^^^^^^^^^^
Error: This variant or record definition does not match that of type u
       Constructors do not match:
         X of bool
       is not compatible with:
         X of int
       The types are not equal.
```
the new error message is clearly an improvement.

Would it be possible to get the second class of suberror messages without the first?

yallop commented:
In the message

```

       Constructors do not match:
         A of 'b
       is not compatible with:
         A of 'a0
```

why is `'b` printed as `'b`, but `'a` printed as `'a0`?

trefis commented:
Disclaimer: I haven't looked at this PR nor the other related ones in details, but:

> Going through the testsuite changes, I feel that in many(74% in fact) cases the error message just becomes more noisy because the information was already here and is only being repeated in a slightly different form. 

We already have that issue in situations such as:
```
# module M : sig type ('a, 'b) bar = A of 'a end = struct type ('a, 'b) bar = A of 'b end;;
Error: Signature mismatch:
       Modules do not match:
         sig type ('a, 'b) bar = A of 'b end
       is not included in
         sig type ('a, 'b) bar = A of 'a end
       Type declarations do not match:
         type ('a, 'b) bar = A of 'b
       is not included in
         type ('a, 'b) bar = A of 'a
       The types for field A are not equal.
```
And indeed this PR would be adding even more noise there.

However, if I instead have the following (which I claim to be more representative of what people actually write):
```
# module M : sig type ('a, 'b) bar = Foo of 'a | Bar of 'b | A of 'a | B of 'b | C of 'a * 'b end = struct type ('a, 'b) bar = Foo of 'a | Bar of 'b | A of 'b | B of 'b | C of 'a * 'b end;;
Error: Signature mismatch:
       ...
       Type declarations do not match:
         type ('a, 'b) bar =
             Foo of 'a
           | Bar of 'b
           | A of 'b
           | B of 'b
           | C of 'a * 'b
       is not included in
         type ('a, 'b) bar =
             Foo of 'a
           | Bar of 'b
           | A of 'a
           | B of 'b
           | C of 'a * 'b
       The types for field A are not equal.
```
We've managed to reduce the noise, yet this PR would still be useful (assuming we also elide the printing of type declarations).

My point being: the noise seems to be an issue not limited to this PR, and perhaps we should think about it separately from this PR?

Octachron commented:
I think the two issues are coupled together: if we suppose that we can elide the previous error message, then the constructor error message may become too elusive by itself. For instance in

```OCaml
# module M : sig
  type ('a, 'b) bar = Foo of 'a | Bar of 'b | A of 'a | B of 'b | C of 'a * 'b
  type 'b foo = A of 'b
end = struct
  type ('a, 'b) bar = Foo of 'a | Bar of 'b | A of 'b | B of 'b | C of 'a * 'b
  type 'a foo = A of 'a
end;;
Error: Signature mismatch:
       ...
       ...

       Constructors do not match:
         A of 'b
       is not compatible with:
         A of 'a
```
it is important to know from which type the constructor comes from. In this case, maybe the error message could be extended to

```OCaml
       Constructors do not match in the definition of ('a,'b) bar :
         A of 'b
       is not compatible with:
         A of 'a  
```
In other words, depending on the presence of elision, the error messages goes from too verbose to not explicit enough.

Octachron commented:
Also one case that is missing in the testsuite would be
```OCaml
module M: sig
  type ('a,'b) t = A of 'a
end = struct
  type ('b, 'a) t = A of 'a
end
```

trefis commented:
> [...] if we suppose that we can elide the previous error message, then the constructor error message may become too elusive by itself.

FTR: I did not mean to say that elision of the outer elements from the trace was the right mechanism (whether here or in general).

Commit:
added tests for printing type Tvars

Commit:
reseting context in constructors and labels methods before printing

MekhrubonT commented:
> why is `'b` printed as `'b`, but `'a` printed as `'a0`?

It was my bug and I fixed it. 

> Also one case that is missing in the testsuite would be

Added. For this case it's printing: 
```
       Constructors do not match:
         A of 'a
       is not compatible with:
         A of 'a
```

Octachron commented:
This last message seems to be a sign that the message needs more context to be locally understandable. Maybe something like:

```
Constructors do not match:
  ('a,'b) t = ...  | A of 'a ...
is not compatible with:
  ('b,'a) t = ... |  A of 'a ...
```
?

lpw25 commented:
> This last message seems to be a sign that the message needs more context to be locally understandable.

I think that the message probably already has sufficient context from the preceding lines:

```ocaml
Error: Signature mismatch:
       Modules do not match:
         sig type ('b, 'a) t = A of 'a end
       is not included in
         sig type ('a, 'b) t = A of 'a end
       Type declarations do not match:
         type ('b, 'a) t = A of 'a
       is not included in
         type ('a, 'b) t = A of 'a
       Constructors do not match:
         A of 'a
       is not compatible with:
         A of 'a
```

In the vast majority of cases the last part gives you sufficient information and in the other cases, like this one, where the message is confusing you can simply scroll up a few lines and the problem becomes clear.

Can I suggest that we tend towards including a full trace of type errors, and then afterwards look at how elements could potentially be elided? The important thing is that the last part of the message is chosen to be the most likely to hold the key information about the error. Extending traces with further narrowing basically always improves this point.

Most of the recent error message work has focused on improving error messages for beginners. I think it is worth pointing out that the kinds of improvement in this PR are useful no matter how experienced you are because they decrease the amount of time it takes to find the useful information in an error message.

Octachron commented on utils/misc.mli:
```diff
@@ -456,6 +456,7 @@ val print_if :
   Format.formatter -> bool ref -> (Format.formatter -> 'a -> unit) -> 'a -> 'a
 (** [print_if ppf flag fmt x] prints [x] with [fmt] on [ppf] if [b] is true. *)
 
+val suffix : int -> string
```
Since `suffix` is only used when printing english ordinal numbers, you should probably export an `ordinal_number: Format.formatter -> int -> unit`  rather than `suffix`.

Octachron commented on typing/includecore.ml:
```diff
@@ -207,11 +227,16 @@ let report_constructor_mismatch first second decl ppf err =
   | Constructor_names (n, name1, name2) ->
-      pr "Constructors number %i have different names, %s and %s."
-        n (Ident.name name1) (Ident.name name2)
+      pr "%i%s constructors have different names, %s and %s."
```
I would really prefer if we avoided starting sentences with a number.

MekhrubonT replied:
Rolled back changes for this message and changes connected to suffix function as well

Octachron commented on testsuite/tests/typing-modules/records_errors_test.ml:
```diff
@@ -0,0 +1,138 @@
+         f0 : unit * unit * unit * float * unit * unit * unit;
+       is not compatible with:
+         f0 : unit * unit * unit * int * unit * unit * unit;
+       The types are not equal.
```
I am a bit surprised by this example. It is quite lengthy, however, this lenght does not highlight the best point of this PR, since there are only two fields (with a long title) rather than a long list of fields. Is there an incoming PR that refines the error in this case?

MekhrubonT replied:
Yes, the next step is to replace equal, moregenal and subtype functions in Ctype by new, which throw exception with error trace instead of returning false (like unify). With error trace it is possible to zoom the error and tell which part exactly is causing mismatch. For the case above additional output for now is:
```
Type unit * unit * unit * float * unit * unit * unit
is not compatible with type
    unit * unit * unit * int * unit * unit * unit
Type float is not compatible with type int
```

You can get more details [here](https://github.com/MekhrubonT/ocaml/pull/2/files)

Octachron commented:
I definitely agree that having a focused error for constructors and record fields is in-principle a good idea that should be useful for many users and not only beginners.

I was probably a bit too focused on the testsuite error messages that are often too pathological to make the extended error messages shine.  The new error message should fare better on more realistic code base.

At the same time, at the presentation level, this new submessage definitively pushes more error messages in my "wall of error" zone. Is there any work planned on a better elision mechanism on your side @lpw25 or should I start to have a look?

lpw25 commented:
> Is there any work planned on a better elision mechanism on your side @lpw25 or should I start to have a look?

There is not work planned on better elision, so I think it would be good for you to have a look. @MekhrubonT has another couple of PRs coming that will add further narrowing, which will improve the errors further but also increase the potential for a "wall of error".

As I said, I'd rather have a "wall of error" that ends in a well focused message than have a short error that is not narrowed down to the cause, so I think it's better to merge the narrowing and then do the elision. However, I see no reason that we can't find something reasonable that gives a good narrow message without lines and lines of unnecessary context.

It might be worth looking at the module level inclusion errors for inspiration since I think they already have some elision stuff.

gasche commented:
I haven't had the time to follow this discussion to the level of details that it deserves, but I'd like to point out that when discussing these kind of error messages, "if it gets too long I just ignore the whole thing" is a pretty common sort of feedback to receive. (We are lucky when people just look at the first two lines, or the last two lines, or both; it seems that many just skip the whole thing.)

Commit:
Rollbacking changes of Constructor_names and Label_names errors message

Octachron commented on typing/printtyp.mli:
```diff
@@ -94,13 +94,17 @@ val type_scheme_max: ?b_reset_names: bool ->
 val extension_constructor:
     Ident.t -> formatter -> extension_constructor -> unit
+val extension_only_constructor:
+    Ident.t -> formatter -> extension_constructor -> unit
```
Last nitpicking: the difference between `extension_constructor` and `extension_only_constructor` should be documented.

MekhrubonT replied:
Added comments with output examples.

Octachron replied:
Perfect, thanks.

Commit:
added documentation for extension_constructor and extension_only_constructor functions in Printtyp; modified Changes file

Octachron reviewed (approved):
Ok, the current state looks good, and the new error submessage can be really useful in the cases where the information was previously completely missing.
